File: retrolambda/src/main/scala/ClassModifier.scala
import java.lang.reflect.{Method, Modifier}

import Implicits._
import org.objectweb.asm.Opcodes._
import org.objectweb.asm.Type._
import org.objectweb.asm._

import scala.collection.mutable.ArrayBuffer

case class MethodContainer(methodName: String, methodDesc: String, interface: String, signature: String)
class ClassModifier(visitor: ClassVisitor, byteCodeVersion: Int = V1_6) extends ClassVisitor(ASM5, visitor) {

  var methodsToOverride: List[MethodContainer] = Nil
  var className: String = _
  type MethodDesc = (String, String)
  var implementedMethods = ArrayBuffer[MethodDesc]()

  override def visit(version: Int, access: Int, name: String, signature: String, superName: String, interfaces: Array[String]) = {
    val isClass = (access & ACC_INTERFACE) == 0
    if(isClass) {
      val defaultMethods = for {
        interface <- interfaces
        cl = interface.getInternalClass
        m @ DefaultMethod() <- cl.getMethods
      } yield MethodContainer(m.getName, Type.getMethodDescriptor(m), interface, signature)
      println(s"Default methods: ${defaultMethods.mkString(",")}")
      methodsToOverride = defaultMethods.toList
      className = name
    }
    super.visit(byteCodeVersion, access, name, signature, superName, interfaces)
  }

  // gotta keep track of overriden default methods
  override def visitMethod(access: Int, name: String, desc: String, signature: String, exceptions: Array[String]): MethodVisitor = {
    implementedMethods += name -> desc
    new InterfaceToHelperRewriter(super.visitMethod(access, name, desc, signature, exceptions))
  }

  override def visitEnd() = {
    def createProxy(l: MethodContainer) = l match {
      case MethodContainer(name, desc, interface, signature) if !implementedMethods.contains(name -> desc) =>
        val tmp = super.visitMethod(ACC_PUBLIC, name, desc, signature, null)
        tmp.visitVarInsn(ALOAD, 0)
        Type.getArgumentTypes(desc).zipWithIndex.foreach{
          case (PrimitiveLoad(instruction), i) => tmp.visitVarInsn(instruction, i + 1) 
          case (_, i) => tmp.visitVarInsn(ALOAD, i + 1)
        }
        tmp.visitMethodInsn(INVOKESTATIC, interface + "helper", name, desc.addParam(interface), false)
        Type.getReturnType(desc) match {
          case ReturnIns(ins)  => tmp.visitInsn(ins)
          case otherwise: Type => tmp.visitInsn(ARETURN)
        }
        tmp.visitMaxs(0, 0)
        tmp.visitEnd()
      case _ => // noop
    }
    methodsToOverride.foreach{ createProxy }
    super.visitEnd()
  }
}

class InterfaceToHelperRewriter(mv: MethodVisitor) extends MethodVisitor(Opcodes.ASM5, mv) {
  override def visitMethodInsn(opcode: Int, owner: String, name: String, desc: String, isInterface: Boolean) = opcode match {
    case INVOKESPECIAL if isInterface =>
      super.visitMethodInsn(INVOKESTATIC, owner + "helper", name, desc.addParam(owner), false)
    case INVOKESTATIC if isInterface =>
      super.visitMethodInsn(INVOKESTATIC, owner + "helper", name + "$static", desc, false)
    case _ =>
      super.visitMethodInsn(opcode, owner, name, desc, isInterface)
  }
}

object DefaultMethod {
  def unapply(m: Method) = !Modifier.isAbstract(m.getModifiers)
}

object InstructorExtractor {
  val (loadMap, returnMap) = {
    val types = List(INT_TYPE,  LONG_TYPE, FLOAT_TYPE, DOUBLE_TYPE, VOID_TYPE)
    val loads = List(ILOAD,     LLOAD,     FLOAD,      DLOAD, -1) // omg a null
    val rets  = List(IRETURN,   LRETURN,   FRETURN,    DRETURN, RETURN)
    val mkMap = types.zip(_: List[Int]).toMap
    mkMap(loads) -> mkMap(rets)
  }
}

class InstructorExtractor(m: Map[Type, Int]) {
  def unapply(f: Type) = m.get(f)
}
object PrimitiveLoad extends InstructorExtractor(InstructorExtractor.loadMap)
object ReturnIns extends InstructorExtractor(InstructorExtractor.returnMap)

File: retrolambda/src/main/java/net/orfjackal/retrolambda/defaultmethods/ClassModifier.java
// Copyright © 2013-2014 Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://www.apache.org/licenses/LICENSE-2.0

package net.orfjackal.retrolambda.defaultmethods;

import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.*;
import java.util.stream.*;

/**
 * Created by arneball on 2014-08-12.
 */
public class ClassModifier extends ClassVisitor implements Opcodes {
    private final int bytecodeVersion;
    private String[] interfaces;

    private Set<VisitedMethod> visitedMethods = new HashSet<>();
    private Set<MethodContainer> defaultMethods = new HashSet<>();

    public ClassModifier(int bytecodeVersion, ClassVisitor cv) {
        super(ASM5, cv);
        this.bytecodeVersion = bytecodeVersion;
    }

    @Override
    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
        boolean isClass = (access & ACC_INTERFACE) == 0;
        if (isClass) {
            defaultMethods = getMethodsToImplement(interfaces, signature);
        }
        this.interfaces = interfaces;
        System.out.println("Class is " + name + ", non abstract " + isClass + ", Interfaces are " + Arrays.toString(interfaces) + "" +
                ", default methods are " + defaultMethods);
        super.visit(bytecodeVersion, access, name, signature, superName, interfaces);
    }

    private static Set<MethodContainer> getMethodsToImplement(String[] interfaces, String sig) {
        return Stream.of(interfaces)
                .flatMap(iff -> getMethodsToImplement(iff, sig).stream())
                .collect(Collectors.toSet());
    }

    private static Set<MethodContainer> getMethodsToImplement(String interfac, String sig) {
        Class<?> ifClass = Helpers.loadClass(interfac);
        Method[] tmp = ifClass.getMethods();
        System.out.println("Interface: " + interfac + ", methods: " + Arrays.toString(tmp));
        return Stream.of(tmp)
                .filter(m -> !Modifier.isAbstract(m.getModifiers()))
                .filter(Helpers::declaringClassBelongsToInputFiles)
                .map(m -> createMethodContainer(m, interfac, sig))
                .collect(Collectors.toSet());
    }

    private static MethodContainer createMethodContainer(Method m, String interfac, String sig) {
        System.out.println("NEED TO CREATE PROXY TO " + interfac + sig + "\n" + m.getDeclaringClass());
        String methodDescriptor = Type.getMethodDescriptor(m);
        String internalName = Type.getInternalName(m.getDeclaringClass());
        return new MethodContainer(m.getName(), methodDescriptor, internalName, sig, getExceptions(m));
    }

    private static String[] getExceptions(Method m) {
        return Stream.of(m.getExceptionTypes())
                .map(Type::getInternalName)
                .toArray(String[]::new);
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
        visitedMethods.add(new VisitedMethod(name, desc));
        return new InterfaceToHelperRewriter(super.visitMethod(access, name, desc, signature, exceptions));
    }

    @Override
    public void visitEnd() {
        for (MethodContainer m : defaultMethods) {
            if (visitedMethods.contains(new VisitedMethod(m.methodName, m.methodDesc))) {
                continue;
            }
            System.out.println("VISITEND, CREATING PROXY " + m);
            MethodVisitor tmp = super.visitMethod(ACC_PUBLIC, m.methodName, m.methodDesc, m.signature, m.exceptions);
            tmp.visitVarInsn(ALOAD, 0);
            int i = 1;
            for (Type arg : Type.getArgumentTypes(m.methodDesc)) {
                tmp.visitVarInsn(arg.getOpcode(ILOAD), i++);
            }
            String rightInterface = findRightInterace(m, interfaces);
            System.out.println("It thinks that the right interface is " + rightInterface);
            String mDesc = Helpers.addParam(m.methodDesc, rightInterface);
            tmp.visitMethodInsn(INVOKESTATIC, rightInterface + "$helper", m.methodName, mDesc, false);
            tmp.visitInsn(Type.getReturnType(m.methodDesc).getOpcode(IRETURN));
            tmp.visitMaxs(0, 0);
            tmp.visitEnd();
        }
        super.visitEnd();
    }

    public static final Comparator<Method> COMPARATOR = (o1, o2) -> {
        boolean o1iso2 = o1.getDeclaringClass().isAssignableFrom(o2.getDeclaringClass());
        boolean o2iso1 = o2.getDeclaringClass().isAssignableFrom(o1.getDeclaringClass());
        if (o1iso2 && o2iso1) {
            return 0;
        } else if (o2iso1) {
            return -1;
        } else {
            return 1;
        }
    };

    public static String findRightInterace(MethodContainer methodContainer, String[] interfaces) {
        System.out.println("Find right interfaces for " + methodContainer + " " + Arrays.toString(interfaces));
        return Stream.of(interfaces)
                .map(Helpers::loadClass)
                .flatMap(i -> flattenInterfaces(i).stream())
                .flatMap(i -> Stream.of(i.getMethods()))
                .filter(Method::isDefault)
                .filter(m -> Type.getMethodDescriptor(m).equals(methodContainer.methodDesc))
                .min(COMPARATOR)
                .map(Method::getDeclaringClass)
                .map(Class::getName)
                .map(s -> s.replace(".", "/"))
                .orElseThrow(NullPointerException::new);
    }

    private static List<Class<?>> flattenInterfaces(Class<?> iff) {
        List<Class<?>> tmp = new ArrayList<>();
        if (iff != null) {
            tmp.add(iff);
            for (Class<?> stream : iff.getInterfaces()) {
                tmp.addAll(flattenInterfaces(stream));
            }
        }
        return tmp;
    }

}